lib/cn/cn_tree_internal.h

```diff
@@ -222,18 +237,18 @@ cn_tree_node_scatter(const struct cn_tree_node *tn);
 void
 cn_comp_slice_cb(struct sts_job *job);
 
-#if HSE_MOCKING
 /**
- * cn_tree_find_node() - Map a node location to a node pointer.
+ * cn_tree_node_find() - Find a cn tree node by node ID.
  *
  * @tree:   tree to search
  * @nodeid: node ID
  *
- * Returns NULL if tree @tree has no node at location @loc.
+ * Return: Node that matches %nodeid or NULL.
  */
 struct cn_tree_node *
-cn_tree_find_node(struct cn_tree *tree, uint64_t nodeid);
+cn_tree_node_find(struct cn_tree *tree, uint64_t nodeid);
```

Comment:
Hierarchical naming is king..

Reply:
We already agreed we were changing this in the public API for 3.0. Let's be consistent

Reply:
I don't recall ever agreeing to any such thing, when did this happen???  The majority of our interfaces are hierarchically named (in increasing order of specificity from left to right), are you suggesting we change them all so that instead of being logically grouped the names are a giant incomprehensible soup?

Reply:
We have a soup right now because nothing is consistent. `cn_get_dataset()` for instance. I would say the majority of our interfaces are not hierarchically named. They are harder to read that way.

```
hse_kvdb_create_kvs()

vs

hse_kvdb_kvs_create()
```

The first one is analogous to the following in pseudocode:

```
class Kvdb:
  static create_kvs()
```

The second one is analogous to the following in my opinion:

```
class Kvs:
  static create()
```

The intention of the function is definitely the first, but the name implies the second. I have already taken the liberty of changing this in the Java bindings to be createKvs() instead of kvsCreate() which matches the convention of the language. More libraries do it the other way in my opinion. Can't even think of a library which does it like us.

Reply:
This conversation occurred early in the 3.0 planning phase.

Reply:
In this example in particular, find_node() is a method of the cn_tree

Reply:
Yeah, that's just a mess.  Any sane naming scheme for a library, module, subsystem, whatever creates a namespace for itself starting with it's name and getting more specific as you go to the right.  Being a library, it would be even better if all our functions started with hse_, but I digress.  Here's a simple two-level example:

https://www.gnu.org.ua/software/gdbm/manual/ndbm.html

Similarly, ADT APIs like list_insert(), list_delete(), ...  From long standing convention it's easy to see we have insert and delete operations wthin the list family.  But what does deleteList() do?  Not to mention looking at stack traces where with hierarchical naming you can easily see the progression through the layers versus the alternative the looks like you're just calling functions that appear to have no relation to one another...  Same with object dumps.

I can search for cn_tree_node_* and find all the operations on cn_tree_nodes if the verbs are at the end.  But if they're not then I have to wade there everything with cn_tree_ in it which just makes it harder to understand and remember the boundaries and capabilities of a "thing".

Sure, sometimes things read a little awkward, but the benefits are well worth it.  I really cannot understand why anyone would not prefer well organized naming to random naming.   Yes, we have not been consistent because different things were written by different people and there was no empowered lead architect looking out for these things since as far as I can remember.   Sigh...

Reply:
Actually, you make a great point about those being tree operations.  I stand corrected, will fix.

But back to my rant!  Think about objects, they are intrinsically hierarchically named.  Namespaces exist to separate one persons pile of garbage from another's, without them it's just a giant bag of functions with no obvious cohesion.  Another example is xxhash, they start all their function names with various versions of XX..., and they even provide a way to rename the prefix during the build to avoid namespace clashes or to hide them from unintended linkage...

Look at kvdb_rest.c.  It's a mix of functions some starting with kvdb_rest_, and curiously, a bunch starting with rest_kvdb_.  That's just sloppy inattention to detail.  In a stack trace I would think that functions starting with "rest_" are part of the rest module, not the kvdb module...

Reply:
I'm changing cn_tree_node_find() to cn_tree_find_node(), cn_tree_node_foreach() to cn_tree_foreach_node(), and so on...
Then I'll see if I can get rid of tree_iter.h

Reply:
Regarding the public/stable API, we did a complete scrub for 2.0 and I have no desire to put users, SQA, or anyone else through wholesales changes for 3.0.  No doubt some API changes will be required to support semantic or functionality enhancements, but we should keep those as narrow as practical.

In any case, this is a higher-level discussion we can have offline.

Reply:
Tristan, I look into removing the tree iter stuff.  There is still one client (cn_metrics.c), so removing it now just means moving a bunch of stuff around just to be able to delete tree_iter.h.  If we wait until we remove cn_metrics.c then we can just simply delete it all.
In a similar vein we have the cn-view and kvset-view stuff which have only one client (kvdb_rest), but like the tree-iter stuff they appear to be set up for extremely limited use by external clients.  cn_metrics.c actually ends up having to pull in cn_tree_internal.h to get the info it needs, so it seems that if we really want to be able to give external clients intimate access to hse internals then we need to spend more time thinking about it.